src/cstruct_ext.ml

```diff
@@ -0,0 +1,20 @@
+(* [or_empty cs] is [cs] unless [Cstruct.is_empty cs] where it is [Cstruct.empty].
+   An empty cstruct will keep a live reference to its underlying buffer. It is
+   preferable to keep a reference live to [Cstruct.empty.buffer] than
+   [cs.buffer] if [cs] is empty. *)
+let[@ocaml.inline always] or_empty cs =
+  if Cstruct.is_empty cs then Cstruct.empty else cs
+
+(* [Cstruct.sub cs 0 cs.len] will result in an empty cstruct **that keeps a
+   live reference to [cs.buffer]**!! *)
+let sub cs off len = or_empty (Cstruct.sub cs off len)
+
+(* [Cstruct.shift cs cs.len] has a similar story *)
+let shift cs len = or_empty (Cstruct.shift cs len)
+
+(* If we append the empty cstruct and we don't need a fresh copy we can do
+   nothing. This has different semantics than Cstruct.append. *)
+let append_nocopy cs cs' =
```

Comment:
hmm hmm, the name is a bit misleading

Reply:
Right, it can't avoid copying in all cases. How about `append'` then?

Reply:
Yes, `append'` sounds good - or "append_avoid_copy_if_empty"? I'd be interested to upstream this into the Cstruct library, but then it modifies the semantics (that some may rely on, esp. in a concurrent setting), and also I try to remove Cstruct from packages ;)

Reply:
I think they might be interesting to upstream. The `append` alternative would have to have a different name due to differing semantics that might matter in some cases. Then again as you say we probably want to use cstruct less.